### PR TITLE
HWKAPM-658 Fix issue with 'source' URI not being calculated correctly…

### PR DESCRIPTION
**Pull request by objectiser**
… in trace completion time processing and communication details deriver

### CODE REVIEW

**Comment by objectiser** [comment]
@pavolloffay Could you please review on your return?

Reviewer notes:
* As zipkin and opentracing primarily use a name (or operation name in zipkin) I thought I would test a scenario with just operation details and no URIs - example data is attached to the jira https://issues.jboss.org/browse/HWKAPM-658
* Should have resulted in the initial client node calling a 'GetBooks' node which then called a 'GetInventory' node, however the result was a tree with the initial client node being the parent of the other two
* This was due to the 'source endpoint' not being initialised correctly - the URI and operation details were being initialised separately, so in this PR have created an 'EndpointRef' class which is now used to hold the details.

I've created a separate jira (659) to look at refactoring this area, but this PR just primarily details with fixing the immediate problem. There is another jira (660) that will discuss how the 'source endpoint' should be calculated - as currently if the fragment does not begin with a Consumer, it will try to find the first Producer and use its details. This is not so necessary when using zipkin/opentracing where the app is more in control of naming of the initial (top level) spans - so may be worth revisiting this approach.

File: api/src/main/java/org/hawkular/apm/api/model/events/EndpointRef.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+public class EndpointRef implements Externalizable {
+
+    private String uri;
+    private String operation;
+    private boolean client = false;
+
+    /**
+     * The default constructor.
+     */
+    public EndpointRef() {
```
Comment (pavolloffay) [inline api/src/main/java/org/hawkular/apm/api/model/events/EndpointRef.java]: Is this class valid with uri and operation null?

File: api/src/main/java/org/hawkular/apm/api/model/events/EndpointRef.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+import org.hawkular.apm.api.model.Constants;
+import org.hawkular.apm.api.utils.EndpointUtil;
+import org.hawkular.apm.api.utils.SerializationUtil;
+
+/**
+ * This class represents an endpoint reference.
+ *
+ * @author gbrown
+ */
+public class EndpointRef implements Externalizable {
```
Comment (pavolloffay) [inline api/src/main/java/org/hawkular/apm/api/model/events/EndpointRef.java]: Why not `Serializable`?

Comment (objectiser) [inline api/src/main/java/org/hawkular/apm/api/model/events/EndpointRef.java]: I've used in other places, so consistent - but probably not such a performance issue as it used to be, so I'll create a separate PR to replace all uses.

File: api/src/main/java/org/hawkular/apm/api/model/events/EndpointRef.java (hunk 1 above)
Comment (objectiser) [inline api/src/main/java/org/hawkular/apm/api/model/events/EndpointRef.java]: It wouldn't be meaningful without the uri or operation.

**Comment by objectiser** [comment]
@pavolloffay Thanks!

File: api/src/main/java/org/hawkular/apm/api/model/events/EndpointRef.java (hunk 1 above)
Comment (pavolloffay) [inline api/src/main/java/org/hawkular/apm/api/model/events/EndpointRef.java]: Then it would be better to don't implement a default constructor.

**Pull request merged**